aptos-move/move-examples/defi/aptoswap/core/sources/liquidity_pool.move

```diff
@@ -0,0 +1,417 @@
+
+    /// Stored at lp token metadata object to find the liquidity pool.
+    /// x address must be lexicographically smaller than y address.
+    struct SwapPair has copy, drop, key {
+        x: Object<Metadata>,
+        y: Object<Metadata>,
+    }
+
+    struct LiquidityPool has key {
+        extend_ref: ExtendRef,
+        store_pair: vector<Object<FungibleStore>>,
+        lp_token_mint_ref: fungible_asset::MintRef,
+        lp_token_burn_ref: fungible_asset::BurnRef,
+        fee: u64,
+        // basis points
+    }
+
+    /// The onchain index of the liquidity pools of different pairs
+    struct LiquidityPoolManager has key {
+        extend_ref: ExtendRef,
+        map: SmartTable<vector<Object<Metadata>>, Object<LiquidityPool>>
+    }
+
+    /// Intialize the module in pool manager object
+    fun init_module(admin: &signer) {
+        let extend_ref = aptoswap::init::retrieve_extend_ref(admin);
+        let pool_manager_obj_signer = object::generate_signer_for_extending(&extend_ref);
+        move_to(&pool_manager_obj_signer, LiquidityPoolManager {
+            extend_ref,
+            map: smart_table::new()
+        })
+    }
+
+    /// Register liquidity pool `X`/`Y`.
+    public fun register(acc: &signer, x: Object<Metadata>, y: Object<Metadata>) acquires LiquidityPoolManager {
+        if (!is_valid_pair(x, y)) {
+            return register(acc, y, x)
+        };
+        assert!(!lp_exists(x, y), error::already_exists(ERR_POOL_EXISTENCE));
+
+        let (lp_token_name, lp_token_symbol) = utilities::swap_lp_token_name_and_symbol(x, y);
+
+        let pool_manager_signer = &get_pool_account_signer();
+        let pool_cref = &object::create_object_from_account(pool_manager_signer);
+        let pool_signer = &object::generate_signer(pool_cref);
+
+        // Create lp fungible token metadata object.
+        let (lp_token_mint_ref, lp_token_burn_ref) = {
+            let seed = b"aptoswap::lp_token_metadata_address";
+            vector::append(&mut seed, bcs::to_bytes(&object::object_address(&x)));
+            vector::append(&mut seed, bcs::to_bytes(&object::object_address(&y)));
+
+            let cref = &object::create_named_object(pool_signer, seed);
+            // Add the pair info to the lp token metadata object
+            move_to(&object::generate_signer(cref), SwapPair { x, y });
+            primary_fungible_store::create_primary_store_enabled_fungible_asset(
+                cref,
+                option::none(),
+                lp_token_name,
+                lp_token_symbol,
+                6,
+                string::utf8(b"http://aptoswap.com/favicon.ico"),
+                string::utf8(b"http://aptoswap.com/")
+            );
+            (
+                fungible_asset::generate_mint_ref(cref),
+                fungible_asset::generate_burn_ref(cref),
+            )
+        };
+
+        let x_store = fungible_asset::create_store(&object::create_object_from_object(pool_signer), x);
```

Comment:
any chance adding docs/diagram to explain the object layout, given readers are not very familiar with these object apis including:

```
let pool_cref = &object::create_object_from_account(pool_manager_signer);
...
let x_store = fungible_asset::create_store(&object::create_object_from_object(pool_signer), x);
```

the main goal is to help users understand what are the different objects and what resources do they hold. it seems to be a bit difficult to reason about the layout compared to the vanilla account model that we're quite familiar with

Reply:
You are right. The object is actually a bag of things instead of one thing. cc @movekevin